Implement Square Pad and Resize Preprocessors

f4str opened the pull request:
# Description

A new module `art.preprocessing.image` was created which will contain all image preprocessors. Under this module, new image preprocessors for square padding and resizing were created. Both of these preprocessors work for both the classification and object detection tasks. For classification, the labels are not changed. For object detection, the bounding boxes of the labels are shifted/scaled accordingly based on the padding or resizing.

The Square Pad preprocessor (`ImageSquarePad`, `ImageSquarePadPyTorch`, and `ImageSquarePadTensorFlowV2`) will pad all images to be a square shape. The user is capable of specifying the type of padding (e.g., constant or symmetric) and any additional parameters taken by the corresponding framework-specific padding function. This preprocessor can also take either a ndarray/tensor of the same image size or a list of ndarray/tensor of arbitrary sizes and still square pad accordingly.

The Resize preprocessor (`ImageResize`, `ImageResizePyTorch`, and `ImageResizeTensorFlowV2`) will resize all image to the be same user specified height and width. The user is capable of specifying the interpolation method used for resizing (e.g., bilinear or area). This preprocessor can also take either a ndarray/tensor of the same image size or a list of ndarray/tensor of arbitrary sizes and still resize all images to the same dimensions. This preprocessor will always convert the resized images into one large ndarray/tensor so this is ideal last step for the training pipeline.

For training, the Square Pad preprocessor should be followed by the Resize processor so images of arbitrary sizes can be padded to preserve their aspect ratio and then resized to be fit into a model. These preprocessors are necessary for properly training the YOLO estimators.

Fixes #2130 
Fixes #2082 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit tests for Square Pad preprocessor
- [x] Unit tests for Resize preprocessor

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

beat-buesser commented on tests/preprocessing/image/test_image_resize.py:
```diff
@@ -0,0 +1,251 @@
+# MIT License
```
I think a new call to pytest for tests/preprocessing/image is needed in run_tests.sh after line 39.

beat-buesser commented on art/preprocessing/image/image_square_pad/tensorflow.py:
```diff
@@ -0,0 +1,161 @@
+        self.verbose = verbose
+        self._check_params()
+
+    def forward(  # type: ignore
```
Are the `type: ignore` comments in the `forward` and `__call__` methods because of the object detection label format?

f4str replied:
Yes, the `type: ignore` is needed due to the object detection labels having a different type. The alternative would be to modify the types of the base preprocessor class which would be too drastic of a change for now.

beat-buesser commented on tests/preprocessing/image/test_image_resize.py:
```diff
@@ -0,0 +1,251 @@
+        data_shape = (2, channels, height, width)
+    else:
+        data_shape = (2, height, width, channels)
+    x = (0.5 * np.ones((data_shape))).astype(ART_NUMPY_DTYPE)
```
Potentially redundant parentheses:
```suggestion
    x = (0.5 * np.ones(data_shape)).astype(ART_NUMPY_DTYPE)
```

beat-buesser commented on tests/preprocessing/image/test_image_square_pad.py:
```diff
@@ -0,0 +1,218 @@
+        data_shape = (2, channels, height, width)
+    else:
+        data_shape = (2, height, width, channels)
+    x = (0.5 * np.ones((data_shape))).astype(ART_NUMPY_DTYPE)
```
Potentially redundant parentheses:
```suggestion
    x = (0.5 * np.ones(data_shape)).astype(ART_NUMPY_DTYPE)
```

beat-buesser reviewed (changes requested):
Hi @f4str Thank you very much for your pull request! I have added a few questions, what do you think?

f4str commented:
Hi @beat-buesser thank you for the comments! I've addressed the review comments and have made the corresponding changes. Feel free to let me know if there any other issues.

beat-buesser reviewed (approved):
Hi @f4str Thank you very much for contributing image preprocessors to ART!
